dm/dm/master/scheduler/scheduler.go

```diff
@@ -693,15 +720,71 @@ func (s *Scheduler) TransferSource(source, worker string) error {
 	if err2 := s.updateStatusToBound(w, ha.NewSourceBound(source, worker)); err2 != nil {
 		s.logger.DPanic("we have checked w.stage is free, so there should not be an error", zap.Error(err2))
 	}
-
-	// 6. try bound the old worker
+	// 6. now this old worker is free, try bound source to it
 	_, err = s.tryBoundForWorker(oldWorker)
 	if err != nil {
 		s.logger.Warn("in transfer source, error when try bound the old worker", zap.Error(err))
 	}
+	s.mu.Unlock()
 	return nil
 }
 
+// batchOperateTaskOnWorker batch operate tasks in one worker and use query-status to make sure all tasks are in expected stage if needWait=true.
+func (s *Scheduler) batchOperateTaskOnWorker(
+	ctx context.Context, worker *Worker, tasks []string, source string, stage pb.Stage, needWait bool) error {
+	for _, taskName := range tasks {
+		if err := s.UpdateExpectSubTaskStage(stage, taskName, source); err != nil {
+			return err
+		}
+	}
+	if !needWait {
+		return nil
+	}
+	// wait all tasks are in expected stage before actually starting scheduling
+WaitLoop:
+	for maxRetryNum := 10; maxRetryNum > 0; maxRetryNum-- {
+		resp, err := worker.queryStatus(ctx)
+		if err != nil {
+			return terror.Annotatef(err, "failed to query worker: %s status", worker.baseInfo.Name)
+		}
+
+		failpoint.Inject("batchOperateTaskOnWorkerMustRetry", func(v failpoint.Value) {
```

Comment:
you call `batchOperateTaskOnWorkerMustRetry` with argument `11` and `8`. What do you expect for them?

Reply:
sorry i don't get your point, i want to check if this wait loop's retry logic, it should retry atmost 10 times 

```
# failed after 10 times retry
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=10]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=0] ["sleep time"=0s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=9]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=1] ["sleep time"=1s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=8]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=2] ["sleep time"=2s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=7]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=3] ["sleep time"=3s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.327 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=6]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=4] ["sleep time"=4s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=5]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=5] ["sleep time"=5s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=4]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=6] ["sleep time"=6s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=3]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=7] ["sleep time"=7s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=2]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=8] ["sleep time"=8s] ["want stage"=Paused] ["current stage"=InvalidStage]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=1]
[2021/12/28 14:31:38.328 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=9] ["sleep time"=9s] ["want stage"=Paused] ["current stage"=InvalidStage]

---
# successed after 2 time retry
[2021/12/28 14:31:38.346 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=10]
[2021/12/28 14:31:38.346 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=0] ["sleep time"=0s] ["want stage"=Paused] ["current stage"=InvalidStage]

[2021/12/28 14:31:38.346 +08:00] [INFO] [scheduler.go:754] ["batchOperateTaskOnWorkerMustRetry failpoint triggered"] [maxRetryNum=9]
[2021/12/28 14:31:38.346 +08:00] [INFO] [scheduler.go:764] ["waiting task"] [component=scheduler] [task=] ["retry times"=1] ["sleep time"=1s] ["want stage"=Paused] ["current stage"=InvalidStage]

[2021/12/28 14:31:38.346 +08:00] [INFO] [scheduler.go:756] ["batchOperateTaskOnWorkerMustRetry passed"] [maxRetryNum=8]
[2021/12/28 14:31:38.363 +08:00] [INFO] [scheduler.go:2070] ["no unbound sources need to bound"] [component=scheduler] [worker="{\"name\":\"dm-worker-4\",\"addr\":\"\"}"]
```

Reply:
I think batchOperateTaskOnWorkerMustRetry=11 means we expect 11 retry, and batchOperateTaskOnWorkerMustRetry=8 means we expect 8 retry.

I prefer
- use `for retry := 0; retry < maxRetryNum; retry++` and `if retry < MustRetry {trigger failpoint}`
- or use `2*return()` syntax https://github.com/pingcap/failpoint